### ISSUE TITLE
crashes with zend_mm_heap corrupted

### ISSUE THREAD

**Issue by ralfbecker** [iuscommunity-pkg/php70u-pecl-memcached#1]
To test I used a Kubernetes cluster running two memcached servers and the following image:
https://quay.io/repository/egroupware/php_ius/image/0a841b570d7cdd5be81dc83dea8552e26291a670d08aaeb6feae2c14a8d13470

I run the following test with memcached session-handler and default files:

```
bash-3.2$ kubectl run test --image=quay.io/egroupware/php_ius:7.0.15 -it --restart=Never
Waiting for pod default/test to be running, status is Pending, pod ready: false
If you don't see a command prompt, try pressing enter.
[root@test /]# php -a
Interactive shell

php > var_dump(ini_get('session.save_handler'));
string(9) "memcached"
php > var_dump(ini_get('session.save_path'));
string(33) "memcached1:11211,memcached2:11211"
php > var_dump(session_start());
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
PHP Warning:  session_start(): Cannot send session cache limiter - headers already sent (output started at php shell code:1) in php shell code on line 1
bool(true)
# The 2 Warnings are expected running the test on cli!
php > $_SESSION['test'] = true;
php > var_dump($_SESSION['test']);
bool(true)
php > var_dump($id = session_id());
string(26) "16qkk8kn469288365p2egg7bk6"
php > session_write_close();
# now reopening the session
php > session_id($id); var_dump(session_start());
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
zend_mm_heap corrupted
[root@test /]#
```
Reopening the session fails and crashes php.  As a test I run the same code with default session-handler files:
```
[root@test /]# php -a
Interactive shell
php > ini_set('session.save_handler', 'files');
php > ini_set('session.save_path', '/tmp');
php > var_dump(session_start());
bool(true)
php > var_dump($id=session_id());
string(26) "i31fegutibh77478jhrb92paj3"
php > $_SESSION['test']=true;
php > session_write_close();
# reopening the session
php > var_dump(session_id($id));
string(26) "i31fegutibh77478jhrb92paj3"
php > var_dump(session_start());
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
PHP Warning:  session_start(): Cannot send session cache limiter - headers already sent (output started at php shell code:1) in php shell code on line 1
bool(true)
php > var_dump($_SESSION['test']);
bool(true)
```
With default session-handler files, sessions work as expected.

I got the same result with a single memcached server, if php-memcached would be - in contrary what it's developer says - affected by the php-memcache (no d!) bug with multiple servers allocating 4GB of ram.

Could it be that lowering/easing the libmemcached dependency caused that?
https://github.com/iuscommunity-pkg/php70u-pecl-memcached/commit/0f1cf6da0f1e27f125fb2bb88418e4e1497d004e
I remember someone mentioned the new libmemcached version was required and that was also what Remi required.

I believe you wanted to build a newer version, thought I could not find it and used the stock one coming with current CentOS7 (1.0.16).

I'm happy to test with a current libmemcached, if you build it.
Or do you have any other idea what I could test or what's going wrong.

Ralf

**Comment by carlwgeorge** [comment]
Lowering the libmemcached requirement was intentional in hopes of using stock libmemcached on EL7.  I discussed this [here](https://github.com/php-memcached-dev/php-memcached/issues/310#issuecomment-278667557) and linked to it from both the [php70u-pecl-memcached](https://github.com/iuscommunity/wishlist/issues/34) and the [php71u-pecl-memcached](https://github.com/iuscommunity/wishlist/issues/111) requests.  I'm under the impression that upstream would like to maintain compatibility with libmemcached 1.0.16.  Is that fair to say @sodabrew?  Would the above be more appropriate as an issue on your project?  Or is it time to finally just set the minimum libmemcached version to 1.0.18 (which I'm fine with)?

**Comment by ralfbecker** [comment]
Setting session.lazy_write = false does NOT fix the issue:

```
bash-3.2$ kubectl run test --image=quay.io/egroupware/php_ius:7.0.15 -it --restart=Never
Waiting for pod default/test to be running, status is Pending, pod ready: false
If you don't see a command prompt, try pressing enter.
[root@test /]# php -a
Interactive shell
php >  ini_set('session.lazy_write', false);
php > session_start();
php > var_dump($id=session_id());
string(26) "rg9t28us4nuev3ukqn2u564lp6"
php > $_SESSION['test']=true;
php > session_write_close();
php > var_dump(session_id($id));
string(26) "rg9t28us4nuev3ukqn2u564lp6"
php > var_dump(session_start());
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
zend_mm_heap corrupted
```

**Comment by sodabrew** [comment]
The minimum libmemcached with the memcached 3.x extension is 1.0.2. I know EL7 has 1.0.16, and that had a particular bug with binary protocol that causes a warning to switch to ascii protocol. Hence the recommendation to use 1.0.18.

**Comment by sodabrew** [comment]
Heap corruption is bad news! Feel free to bring this bug report upstream to php-memcached if needed.

**Comment by ralfbecker** [comment]
@sodabrew Sure I can post it at php-memcached too, question if that makes sense at the current state?

My initial post contained our memcached session configuration in a link:
```
session.save_handler = memcached
session.save_path = "memcached1:11211,memcached2:11211"
session.gc_maxlifetime = 14400
session.serialize_handler = igbinary
memcached.sess_persistent = On
memcached.sess_number_of_replicas = 2
memcached.sess_connect_timeout = 50
memcached.sess_lock_wait_min = 250
memcached.sess_lock_wait_max = 2000
memcached.sess_lock_retries = 20
```
(last 3 entries are used as we have many ajax requests, so we want an initial small timeout, but we also made the experience (with old php-memcache extension) that it's default lock-timeout of 15s, which is close to yours of 20s, caused not verifyable sessions, which a higher lock-timeout fixed)

I'm happy to test with different settings, thought I don't want to test all permutations.

**Comment by ralfbecker** [comment]
Following config works (listing only changes from above not working config):
```
session.save_path = "memcached1:11211"
session.serialize_handler = php
memcached.sess_persistent = Off
memcached.sess_number_of_replicas = 0
```

So it's either the 2 memcached servers, or persistent connection or the serialize handler ;-)

**Comment by ralfbecker** [comment]
Using 2 memcached servers (listing only changes from above):
```
session.save_path = "memcached1:11211,memcached2:11211"
memcached.sess_number_of_replicas = 2
```

Still works too:
```
[root@mserver-2036111592-1z87c /]# php -a
Interactive shell
php > session_start();
php > var_dump($id=session_id());
string(26) "0p6dd2bh4p3mb3sgvm7a7mdh52"
php > $_SESSION['test']=true;
php > session_write_close();
php > var_dump(session_id($id));
string(26) "0p6dd2bh4p3mb3sgvm7a7mdh52"
php > var_dump(session_start());
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
PHP Warning:  session_start(): Cannot send session cache limiter - headers already sent (output started at php shell code:1) in php shell code on line 1
bool(true)
php > var_dump($_SESSION);
array(1) {
  ["test"]=>
  bool(true)
}
```

Down to either persisten connection or session.serialize_handler = igbinary ...

**Comment by ralfbecker** [comment]
Seems to be the persistent connection:
```
memcached.sess_persistent = On
```
```
[root@mserver-2036111592-1z87c /]# php -a
Interactive shell
php > session_start();
php > var_dump($id=session_id());
string(26) "mjbk9fph9tj0t28pg55sskcli4"
php > $_SESSION['test']=true;
php > session_write_close();
php > var_dump(session_id($id));
string(26) "mjbk9fph9tj0t28pg55sskcli4"
php > session_start();
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
zend_mm_heap corrupted
```

--> testing my original config without the persistent connection

**Comment by ralfbecker** [comment]
Ok, it comes down to persistent connection:
```
memcached.sess_persistent = Off
session.serialize_handler = igbinary
```
```
[root@mserver-2036111592-1z87c /]# php -a
Interactive shell
php > session_start();
php > var_dump($id=session_id());
string(26) "p34sndhhiv6r67j5jndi6a0mf6"
php > $_SESSION['test']=true;
php > session_write_close();
php > var_dump(session_id($id));
string(26) "p34sndhhiv6r67j5jndi6a0mf6"
php > session_start();
PHP Warning:  session_start(): Cannot send session cookie - headers already sent by (output started at php shell code:1) in php shell code on line 1
PHP Warning:  session_start(): Cannot send session cache limiter - headers already sent (output started at php shell code:1) in php shell code on line 1
php > var_dump($_SESSION);
array(1) {
  ["test"]=>
  bool(true)
}
```
--> so persistent connection seems to cause the zend_mm_heap corrupted (maybe in combination with any of the above, I'll try with only one server, so basically default config with just persistent connections on)

**Comment by ralfbecker** [comment]
An other observation ```session.lazy_write = On```, which is the default (and not in IUS /etc/php.ini!) gives terrible performance with libmemcached 1.0.16, seems like some kind of timeout.
@carlwgeorge if you keep going with libmemcached 1.0.16, this only works reasonable with session.lazy_write off. Maybe put that in /etc/php.d/50-memcached.ini, if that's possible ...

**Comment by ralfbecker** [comment]
@sodabrew As a final test I tried with a single memcached:
```
[root@mserver-2036111592-1z87c /]# php -a
Interactive shell
php > ini_set('memcached.sess_persistent', true);
php > ini_set('session.save_path', 'memcached1:11211');
php > ini_set('memcached.sess_number_of_replicas', 0);
php > session_start(); $id=session_id(); $_SESSION['test']=true; session_write_close();
php >  session_id($id); session_start(); var_dump($_SESSION); session_write_close();
zend_mm_heap corrupted
```

So I conclude:
1.) ```memcached.sess_persistent = 1``` is causing zend_mm_heap corrupted in 3.0.1 (at least with libmemcached 1.0.16)
2.) ```session.lazy_write = On``` with libmemcache 1.0.16 (at least) results in terrible performance (probably a timeout somewhere)

@sodabrew please let me know if you need more information

**Comment by b-harper** [comment]
@ralfbecker,  Thanks for providing your findings.  I edited your posts just to add some formatting so they are easier to read.

**Comment by carlwgeorge** [comment]
@ralfbecker When you file this issue upstream, can you link it here so I can subscribe to it?

**Comment by sodabrew** [comment]
Great information! There is an upstream bug describing a known issue with libmemcached 1.0.16 and the binary protocol for _timeouts_. Upstream bug report php-memcached-dev/php-memcached#310 has further details.

There's another upstream report about segfaults at php-memcached-dev/php-memcached#265 that seemed to point at special characters being the proximate cause of the problem, but I don't have a complete backtrace yet.

Getting a C-level backtrace with all of the symbols will be super helpful if you can post one here.

**Comment by carlwgeorge** [comment]
@ralfbecker I believe @sodabrew would like you to install php70u-pecl-memcached-debuginfo to get those additional symbols.

**Comment by ralfbecker** [comment]
@sodabrew I saw the upstream bug https://github.com/php-memcached-dev/php-memcached/issues/310. That's why I tried switching session.lazy_write = Off and I can confirm it works. I'm happy to test, if @carlwgeorge can build a newer libmemcached version.

I don't think my problem is related to special chars in the session, as there was no such thin in my examples
```
$_SESSION['test'] = true;
```
 
What do I need to do to get a C-level backtrace, apart from installing php70u-pecl-memcached-debuginfo?

**Comment by sodabrew** [comment]
Once you have the debug symbols, you'll need a way to capture the crash itself. That might be enabling core files, or looking to your web server process (nginx, apache, fpm, etc.) to see if there are options to capture backtraces from child processes that exit with a segfault.

**Comment by ralfbecker** [comment]
I used the cli (php -a) to reproduce it, so not web-server or fpm involved. 

@sodabrew Could I just use something like:
```
gdb php -a
```
and reproduce the bug, to get a backtrace?

**Comment by sodabrew** [comment]
Yes, gdb will work.

**Comment by ralfbecker** [comment]
@sodabrew unfortunately it did not :(
```
[root@test /]# gdb --args php -a
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-94.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /usr/bin/php...Reading symbols from /usr/lib/debug/usr/bin/php.debug...done.
done.
(gdb) run
Starting program: /usr/bin/php -a
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Interactive shell

php > ini_set('memcached.sess_persistent', true); ini_set('session.save_path', 'memcached1:11211'); ini_set('memcached.sess_number_of_replicas', 0);
php > session_start(); $id=session_id(); $_SESSION['test']=true; session_write_close();
php > session_id($id); session_start(); var_dump($_SESSION); session_write_close();
zend_mm_heap corrupted
[Inferior 1 (process 217) exited with code 01]
(gdb) backtrace
No stack.
(gdb)
```

**Comment by carlwgeorge** [comment]
I was able to reproduce this bug on Fedora 25 using libmemcached 1.0.18.  I reported it [upstream](https://github.com/php-memcached-dev/php-memcached/issues/327), so I'm closing this issue out as it doesn't appear to be packaging related (we can reopen later if needed).

**Comment by carlwgeorge** [comment]
FYI @ralfbecker, upstream fixed this in 3.0.3.  I updated our spec file to the new version and new packages will move to our testing repositories during tonight's automation.
